src/bidiMapper/utils/deferred.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2021 Google LLC.
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export class Deferred<T> implements Promise<T> {
+  #resolve: (value: T) => void = () => {};
+  #reject: (value: T) => void = () => {};
+  #promise: Promise<T>;
+
+  constructor() {
+    this.#promise = new Promise((resolve, reject) => {
+      this.#resolve = resolve;
+      this.#reject = reject;
+    });
+  }
+
+  then<TResult1 = T, TResult2 = never>(
+    onFulfilled?:
+      | ((value: T) => TResult1 | PromiseLike<TResult1>)
+      | undefined
```

Comment:
Since `onFulfilled?` is optional, you should not need `| undefined` here. Same applies below (3 more times, including `catch` and `finally`).

Reply:
The original interface was copied from the TypeScript Promise interface definition. But it can be simplified. Done.